apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,726 @@
+        panel: newPanel.name,
+        args
+      } : null,
+      prev: currentState ? {
+        panel: currentState.panel.name,
+        args: currentState.args
+      } : null
+    }
+  };
+
+  Object.assign(navigationTransition.transitionArgs, args);
+
+  navigationTransition.states = {
+    from: currentState,
+    to: {
+      panel: newPanel,
+      args
+    }
+  };
+
+  return Promise.resolve();
+}
+
+/**
+ * Sets the currentState variable to its new value.
+ */
+function setFutureState() {
+  currentState = navigationTransition.states.to;
+}
+
+/**
+ * Ends the navigation so that new navigations can take place.
+ * This cleans up the navigation states.
+ */
+function endNavigation() {
+  debug('endNavigation()');
+  navigationTransition && navigationTransition.defer.resolve();
+  navigationTransition = null;
+}
 
-// This function is used as a rejection handler, to swallow errors and log them
-function catchError(e) {
+/**
+ * This function is used to catch errors at the end of navigation promise
+ * chains. We're ending the navigation if this happens, so that we can start a
+ * new one afterwards.
+ *
+ * @param {Exception} [e] The exception to display, if any.
+ */
+function onNavigationError(e) {
+  console.error('Navigation error', e);
+  endNavigation();
+}
+
+/**
+ * This function is used as a rejection handler, to swallow errors and log them.
+ *
+ * @param {Exception} [e] The exception to display, if any.
+ */
+function catchStepError(e) {
   if (e) {
     console.error(
       'Navigation.toPanel: got an exception: (%s) %s',
       e, e.stack && e.stack.replace(/\n/g, '|')
     );
   } else {
-    console.log('Navigation.toPanel: got a rejected promise');
+    debug('Navigation.toPanel: got a rejected promise');
   }
 }
 
-var currentPanel;
-var queuedPanel;
+/**
+ * Executes the lifecycle function for `stepName` on the right object.
+ *
+ * @param {String} stepName The step to execute: one of 'beforeLeave',
+ * 'beforeEnter', 'afterLeave', 'afterEnter'.
+ * @returns {Promise} Resolved when the step returns a value or a resolved
+ * promise, rejected when the step either throws or returns a rejected promise.
+ * Or a resolved promise if no step was executed (because it was already
+ * executed or there was no function to execute).
+ */
+function executeNavigationStep(stepName) {
+  debug(`executeNavigationStep(${stepName})`);
+  if (!navigationTransition) {
+    throw new Error(
+      'Trying to execute the navigation step ' + stepName +
+      ' but no transition is started yet.'
+    );
+  }
 
-function nextQueuedPanel() {
-  if (queuedPanel) {
-    var next = queuedPanel;
-    queuedPanel = null;
-    Navigation.toPanel(next.panel, next.args)
-      .then(next.defer.resolve, next.defer.reject);
+  if (navigationTransition.stepsRun.indexOf(stepName) >= 0) {
+    // this step has already run (case of back -> onHashChange)
+    return Promise.resolve();
+  }
+
+  var stateKey = stepName.endsWith('Leave') ? 'from' : 'to';
```

Comment:
> convention vs configuration ? ;)

Exactly! I thought that it maybe clearer that we have only predefined list of steps (not ```almostRigtheBeforeLeave``` :)), but don't have strong opinion as well, so choose whatever you want :)